build_runner/lib/src/generate/build_definition.dart

```diff
@@ -271,16 +271,28 @@ class _Loader {
 
   /// Returns the set of original package inputs on disk.
   Future<Set<AssetId>> _findInputSources() async {
-    var inputSets = _options.packageGraph.allPackages.values.map((package) =>
-        new InputSet(package.name, package.includes,
-            excludes: package.excludes));
+    var inputSets =
+        _options.packageGraph.allPackages.values.map(_inputSetForPackage);
     var sources = (await _listAssetIds(inputSets).toSet())
       ..addAll(await _options.reader
           .findAssets(new Glob('$entryPointDir/**'))
           .toSet());
     return sources;
   }
 
+  /// Only allow reads into `lib/` for non-root packages.
+  ///
+  /// As an optimization, also limit the SDK package to the files necessary for
+  /// DDC.
+  InputSet _inputSetForPackage(PackageNode package) {
+    List<String> includes = package.name == r'$sdk'
```

Comment:
Lets move `$sdk` to a constant somewhere, and consider exposing it (build_compilers uses it as well).

I am fine with this being a followup though if you prefer.